src/API/Site/Controllers/MerchantCenter/PriceBenchmarksController.php

```diff
@@ -0,0 +1,231 @@
+					'callback' => $this->get_price_benchmarks_callback(),
+				],
+				'schema' => $this->get_api_response_schema_callback(),
+			]
+		);
+	}
+
+	/**
+	 * Maps query arguments from the REST request.
+	 *
+	 * @param Request $request REST Request.
+	 * @return array
+	 */
+	protected function prepare_query_arguments( Request $request ): array {
+		$args = wp_parse_args(
+			array_intersect_key(
+				$request->get_query_params(),
+				$this->get_collection_params()
+			),
+			$request->get_default_params()
+		);
+
+		return $args;
+	}
+
+	/**
+	 * Get the callback function for the price benchmarks request.
+	 *
+	 * @return callable
+	 */
+	protected function get_price_benchmarks_callback(): callable {
+		return function ( Request $request ) {
+			try {
+				/** @var MerchantPriceBenchmarks $merchant */
+				$merchant = $this->container->get( MerchantPriceBenchmarks::class );
+
+				$benchmark_data      = $merchant->get_benchmark_data( $this->prepare_query_arguments( $request ) );
+				$price_insights_data = $merchant->get_price_insights( $this->prepare_query_arguments( $request ) );
+
+				// Map the data to the required format.
+				$response_data = $this->map_price_benchmarks_response( $benchmark_data, $price_insights_data );
+
+				return new Response( $response_data );
+			} catch ( Exception $e ) {
+				return $this->response_from_exception( $e );
+			}
+		};
+	}
+
+	/**
+	 * Maps the benchmark and price insights data to the required API response format.
+	 *
+	 * @param array $benchmark_data Raw benchmark data.
+	 * @param array $price_insights_data Raw price insights data.
+	 * @return array Mapped response data.
+	 */
+	protected function map_price_benchmarks_response( array $benchmark_data, array $price_insights_data ): array {
+		$mapped_data = [];
+
+		if ( empty( $benchmark_data['results'] ) || empty( $price_insights_data['results'] ) ) {
+			return $mapped_data;
+		}
+		// Loop through the benchmark data results.
+		foreach ( $benchmark_data['results'] as $benchmark_result ) {
+			$product_view          = $benchmark_result['productView'];
+			$price_competitiveness = $benchmark_result['priceCompetitiveness'];
+
+			// Find the matching price insights data by offer_id.
+			foreach ( $price_insights_data['results'] as $price_insights_result ) {
+				if ( $product_view['id'] !== $price_insights_result['productView']['id'] ) {
+					continue;
+				}
+
+				$price_insights = $price_insights_result['priceCompetitiveness'] ?? [];
+
+				// Calculate the price gap.
+				$regular_price   = (int) $product_view['priceMicros'];
+				$price_on_google = (int) $price_competitiveness['benchmarkPriceMicros'];
+				$price_gap       = $regular_price - $price_on_google;
+
+				// Get the WooCommerce product ID and thumbnail.
+				$product_id = $this->product_helper->get_wc_product_id( $product_view['id'] );
+				$thumbnail  = $this->get_product_thumbnail( $product_id );
+
+				// Map the data to the required format.
+				$mapped_data[] = [
```

Comment:
After implementing the previous suggested change to the nested loops, this will need to be updated so that we're looping through the mapped data to return the expected responses.

```
$response = array_map( function ( $item ) { // etc. }, $mapped_data );
```

Reply:
Update in https://github.com/woocommerce/google-listings-and-ads/pull/2838/commits/d6fd66dc2b3f1cf76e99fe0cb404689796b0c7fc